### PR TITLE
Joysticks: fix easter egg

### PR DESCRIPTION
**Pull request by lrusak**
not sure if @garbear has other plans for this but it was lost in the last major refactor.

This makes it work again.

shhhhhhhhhh

### CODE REVIEW

**Comment by garbear** [comment]
Thanks! There's only one change I want to make.

The new system added a level of complexity because I wanted game hotkeys (defined in joystick.xml's `<FullscreenGame>` tag) to work in the game OSD. To solve this, I added a "keymap environment" parameter. Outside of games, the environment's job is to translate the key based on the active window ID. Inside games, I register an additional keymapper, and use the environment parameter to force `<FullscreenGame>` keys.

We only need the top-level keymapper to monitor for easter eggs, so I'll add a variable to the keymap environment to prevent easter egg monitoring in game OSD menus. PR is coming your way.

File: xbmc/input/joysticks/keymaps/KeymapHandler.h [hunk 1]
```diff
@@ -19,6 +19,7 @@
  */
 #pragma once
 
+#include "input/joysticks/IButtonSequence.h"
```
Comment (garbear) [inline xbmc/input/joysticks/keymaps/KeymapHandler.h]: You can remove this include and instead forward-declare `IButtonSequence`

File: xbmc/input/joysticks/keymaps/KeymapHandler.h [hunk 2]
```diff
@@ -63,6 +64,14 @@ namespace JOYSTICK
     virtual bool OnAnalogStickMotion(const FeatureName& feature, float x, float y, unsigned int motionTimeMs) override;
     virtual bool OnAccelerometerMotion(const FeatureName& feature, float x, float y, float z) override;
 
+  protected:
+    /*!
+     * \brief Keep track of cheat code presses
+     *
+     * Child classes should initialize this with the appropriate controller ID.
```
Comment (garbear) [inline xbmc/input/joysticks/keymaps/KeymapHandler.h]: This comment doesn't apply to the new system

File: xbmc/input/joysticks/keymaps/KeymapHandler.h (hunk 1 above)
Comment (lrusak) [inline xbmc/input/joysticks/keymaps/KeymapHandler.h]: without it I get
```
xbmc/input/joysticks/keymaps/KeymapHandler.cpp: In constructor 'KODI::JOYSTICK::CKeymapHandler::CKeymapHandler(IActionListener*, const IKeymap*)':
xbmc/input/joysticks/keymaps/KeymapHandler.cpp:48:59: error: no matching function for call to 'std::unique_ptr<IButtonSequence>::reset(KODI::JOYSTICK::CJoystickEasterEgg*)'
   m_easterEgg.reset(new CJoystickEasterEgg(ControllerID()));
```

Comment (garbear) [inline xbmc/input/joysticks/keymaps/KeymapHandler.h]: Try moving the `CKeymapHandler` destructor to the .cpp file:

```
CKeymapHandler::~CKeymapHandler() = default;
```

Comment (lrusak) [inline xbmc/input/joysticks/keymaps/KeymapHandler.h]: nope

Comment (garbear) [inline xbmc/input/joysticks/keymaps/KeymapHandler.h]: ok, just leave the header then

**Comment by lrusak** [comment]
rebased and squashed

**Comment by garbear** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by da-anda** [comment]
is it still worth having the easter egg feature? Isn't that cat out of the bag already?

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
> is it still worth having the easter egg feature? Isn't that cat out of the bag already?

Do you have any idea how much technical debt I take on the second games are enabled by default? If we hit the switch today I throw away easily the next six months of my life, likely more. I'm not saying I didn't sign up for that, just that I'm not ready to commitment to this yet.

**Comment by lrusak** [comment]
Not only that but we have to solve the infrastructure issue before we can go full on

**Pull request merged**